applications/crossbar/src/modules/cb_storage.erl

```diff
@@ -412,3 +418,84 @@ doc_id({'user', UserId, _AccountId}) -> UserId;
+                                                    orelse ReqVerb =:= ?HTTP_POST
+                                                    orelse ReqVerb =:= ?HTTP_PATCH ->
+    case cb_context:resp_status(Context) of
+        'success' ->
+            lager:debug("Validating storage settings"),
+            Attachments = kz_json:get_json_value(<<"attachments">>,
+                                                 cb_context:doc(Context)),
+            validate_attachments_settings(Attachments, Context);
+        _ ->
+            Context
+    end;
+maybe_check_storage_settings(Context, _ReqVerb) ->
+    Context.
+
+-spec validate_attachments_settings(kz_json:object()
+                                   ,cb_context:context()
+                                   ) -> cb_context:context().
+validate_attachments_settings(Attachments, Context) ->
+    kz_json:foldl(fun validate_attachment_settings_fold/3, Context, Attachments).
+
+-spec validate_attachment_settings_fold(kz_term:ne_binary()
+                                       ,kz_json:object()
+                                       ,cb_context:context()
+                                       ) -> cb_context:context().
+validate_attachment_settings_fold(AttId, Att, ContextAcc) ->
+    %% Files content will differ at least on this value and also this `Random'
+    %% value is used to make sure we always send unique attachment names when
+    %% testing storage settings.
+    Random = kz_binary:rand_hex(16),
+    Content = <<"some random content: ", Random/binary>>,
+    AName = <<Random/binary, "_test_credentials_file.txt">>,
+    DbName = kz_util:format_account_db(cb_context:account_id(ContextAcc)),
+    DocId = doc_id(ContextAcc),
+    Handler = kz_json:get_value(<<"handler">>, Att),
+    Settings = kz_json:get_value(<<"settings">>, Att),
+    AttHandler = kz_term:to_atom(<<"kz_att_", Handler/binary>>, 'true'),
+    AttSettings = kz_maps:keys_to_atoms(kz_json:to_map(Settings)),
+    Opts = [{'plan_override', #{'att_handler' => {AttHandler, AttSettings}
+                               ,'att_post_handler' => 'external'
+                               }}],
+    %% Check the storage settings have permissions to create files
+    case kz_datamgr:put_attachment(DbName, DocId, AName, Content, Opts) of
+        {'ok', _CreatedDoc, _CreatedProps} ->
+            %% Check the storage settings have permissions to read files
+            case kz_datamgr:fetch_attachment(DbName, DocId, AName, Opts) of
+                {'ok', _Content} ->
+                    ContextAcc;
+                Error ->
+                    add_att_settings_validation_error(AttId, Error, ContextAcc)
+            end;
+        Error ->
+            add_att_settings_validation_error(AttId, Error, ContextAcc)
+    end.
+
+-spec add_att_settings_validation_error(kz_term:ne_binary()
+                                       ,gen_attachment:error_response() |
+                                        kz_datamgr:data_error()
```

Comment:
I think this is one of the areas where its confusing having an "opaque" type OR a regular type since obviously gen_attachment:error_code/2 won't work for `{'error', 'invalid_db_name'}` or other kz_datamgr errors.

Reply:
```erlang
cb_storage.erl:473: The call cb_storage:add_att_settings_validation_error(AttId::any(),Error::{'error',atom() | integer() | {'EXIT' | 'conn_failed' | 'url_parsing_failed' | integer(),_} | {'ok',string(),_,_}},ContextAcc::#cb_context{content_types_provided::[{atom(),[atom() | {_,_}]}],content_types_accepted::[{atom(),[atom() | {_,_}]}],allowed_methods::[<<_:8,_:_*8>>],allow_methods::[<<_:8,_:_*8>>],languages_provided::[<<_:8,_:_*8>>],charsets_provided::[<<_:8,_:_*8>>],encodings_provided::[<<_:8,_:_*8>>],auth_token::'undefined' | <<_:8,_:_*8>>,auth_token_type::'basic' | 'oauth' | 'unknown' | 'x-auth-token',auth_account_id::'undefined' | <<_:8,_:_*8>>,auth_doc::'undefined' | {[{atom() | <<_:8,_:_*8>>,atom() | binary() | [any()] | number() | {[{_,_}]}}]},req_verb::<<_:8,_:_*8>>,req_nouns::[{<<_:8,_:_*8>>,[<<_:8,_:_*8>>]}],req_json::{[{atom() | <<_:8,_:_*8>>,atom() | binary() | [any()] | number() | {[{_,_}]}}]} | {'malformed',binary()},req_files::[{<<_:8,_:_*8>>,{[{atom() | <<_:8,_:_*8>>,atom() | binary() | [any()] | number() | {[{_,_}]}}]}}],req_data::atom() | binary() | [any()] | number() | {[{atom() | <<_:8,_:_*8>>,_}]},query_json::{[{atom() | <<_:8,_:_*8>>,atom() | binary() | [any()] | number() | {[{_,_}]}}]},account_id::'undefined' | <<_:8,_:_*8>>,account_name::'undefined' | <<_:8,_:_*8>>,user_id::'undefined' | <<_:8,_:_*8>>,device_id::'undefined' | <<_:8,_:_*8>>,reseller_id::'undefined' | <<_:8,_:_*8>>,db_name::'undefined' | binary() | [<<_:8,_:_*8>>],doc::'undefined' | [{[{atom() | <<_:8,_:_*8>>,atom() | binary() | [any()] | number() | {[{_,_}]}}]}] | {[{atom() | <<_:8,_:_*8>>,atom() | binary() | [any()] | number() | {[{_,_}]}}]},pretty_print::boolean(),resp_expires::{{non_neg_integer(),1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12,1..255},{byte(),byte(),byte()}},resp_etag::'automatic' | 'undefined' | binary() | string(),resp_status::'error' | 'fatal' | 'stop' | 'success',resp_error_msg::'undefined' | <<_:8,_:_*8>>,resp_error_code::'undefined' | integer(),resp_file::binary(),resp_data::atom() | binary() | [any()] | number() | {[{atom() | <<_:8,_:_*8>>,_}]} | map(),resp_envelope::{[{atom() | <<_:8,_:_*8>>,atom() | binary() | [any()] | number() | {[{_,_}]}}]},start::{non_neg_integer(),non_neg_integer(),non_neg_integer()},req_id::binary(),storage::[atom() | {_,_}],raw_host::binary(),port::integer(),raw_path::binary(),raw_qs::binary(),method::<<_:8,_:_*8>>,validation_errors::{[{atom() | <<_:8,_:_*8>>,atom() | binary() | [any()] | number() | {[{_,_}]}}]},client_ip::'undefined' | <<_:8,_:_*8>>,load_merge_bypass::'undefined' | {[{atom() | <<_:8,_:_*8>>,atom() | binary() | [any()] | number() | {[{_,_}]}}]},profile_id::'undefined' | <<_:8,_:_*8>>,api_version::<<_:8,_:_*8>>,magic_pathed::boolean(),should_paginate::'false' | 'true' | 'undefined',host_url::binary()}) does not have an opaque term of type gen_attachment:error_response() as 2nd argument 
```
Dialyzer doesn't like this arrangement either.

Reply:
```erlang
cb_storage.erl:485: The call gen_attachment:error_code(ErrorResponse::{'error',atom() | integer() | {'EXIT' | 'conn_failed' | 'url_parsing_failed' | integer(),_} | {'ok',[any()],_,_}} | {'ok',{[any()]}}) does not have a term of type {'error',[atom() | {_,_}]} | gen_attachment:error_response() (with opaque subterms) as 1st argument
```